MsCorePkg/MacAddressEmulationDxe/MacAddressEmulationDxe.c

```diff
@@ -0,0 +1,239 @@
+/** @file
+
+  DXE Driver for handling MAC Address Emulation.
+
+  Copyright (C) Microsoft Corporation.
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+#include "MacAddressEmulationDxe.h"
+
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/DebugLib.h>
+#include <Library/UefiLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/NetLib.h>
+
+#include "Library/MacAddressEmulationPlatformLib.h"
+
+
+BOOLEAN
+SnpSupportsMacEmuCheck (
+  IN CONST EFI_HANDLE SnpHandle,
+  IN CONST EFI_SIMPLE_NETWORK_PROTOCOL *Snp,
+  IN CONST VOID *SnpContext
+  )
+{
+  MAC_EMULATION_SNP_NOTIFY_CONTEXT* Context;
+  BOOLEAN IsMatch = FALSE;
+
+  if (SnpHandle != NULL && Snp != NULL && SnpContext != NULL) {
+    IsMatch = TRUE;
+    Context = (MAC_EMULATION_SNP_NOTIFY_CONTEXT*) SnpContext;
+  }
+  
+  if (IsMatch && Snp->Mode->State != EfiSimpleNetworkInitialized) {
+    DEBUG ((DEBUG_ERROR, "[%a]: SNP handle in unexpected state %d, cannot update MAC.\n", __FUNCTION__, Snp->Mode->State));
+    IsMatch = FALSE;
+  }
+
+  if (IsMatch && Snp->Mode->IfType != NET_IFTYPE_ETHERNET) {
+    DEBUG ((DEBUG_WARN, "[%a]: SNP interface type is not Ethernet.\n", __FUNCTION__));
+    IsMatch = FALSE;
+  }
+
+  if (IsMatch && !Snp->Mode->MacAddressChangeable) {
+    DEBUG ((DEBUG_WARN, "[%a]: SNP interface does not support MAC address programming", __FUNCTION__));
+    IsMatch = FALSE;
+  }
+
+  if (IsMatch && !SnpSupportsMacEmulation(SnpHandle)) { 
+    IsMatch = FALSE;
+  }
+
+  if (IsMatch && Context->Assigned == TRUE) {
+    // If emulation was already assigned, make sure that this is the same interface that was assigned previously
+    // by comparing the permanent MAC address against the address cached during the first assignment (updated in
+    // context below).
+    IsMatch = (CompareMem (&Snp->Mode->PermanentAddress, &Context->PermanentAddress, NET_ETHER_ADDR_LEN) == 0);
+  }
+
+  return IsMatch;
+}
+
+/**
+  
+**/
+EFI_SIMPLE_NETWORK_PROTOCOL*
+FindMatchingSnp (
+  IN SNP_MATCH_FUNCTION MatchFunction,
```

Comment:
Originally intended to make unit testing this function easier since you can't mock functions within the same .c file, didn't actually need to use it. If the IsMatch function becomes more complex then using a function pointer makes stubbing the behavior much easier.

Reply:
If it's not being used and there are currently no plans to use, can we remove it?  I'm all for keeping things as simple as possible especially since some programmers have trouble with pointers.